File: lib/download.js
'use strict';

const Fs = require('fs');
const ProgressBar = require('progress');
const {PassThrough} = require('stream');
const {once} = require('events');

const Cse = require('./aws/cse');
const S3 = require('./aws/s3');

const internals = {};


exports.download = async function(filePath, bucketName, s3Key, cseKeyArn) {
  const {ContentLength} = await S3.headObject(bucketName, s3Key);

  const progress = new ProgressBar('  downloading [:bar] :percent :etas', {
    complete: '=',
    incomplete: ' ',
    width: 20,
    total: ContentLength
  });


  const writeStream = Fs.createWriteStream(filePath)

  const passThrough = new PassThrough();
  const s3Stream = internals.getObjectStream(bucketName, s3Key, cseKeyArn, passThrough);
  const fileStream = s3Stream.pipe(writeStream);

  passThrough.on('data', function (chunk) {
    progress.tick(chunk.length);
  });

  await once(fileStream, 'finish');
};

internals.getObjectStream = function(bucketName, s3Key, cseKeyArn, passThrough) {
  const s3Stream = S3.getObjectStream(bucketName, s3Key).pipe(passThrough);

  if(cseKeyArn) {
   return s3Stream.pipe(Cse.decryptStream(cseKeyArn));
  }

  return s3Stream;
};

File: lib/download.py
import lib.aws.s3 as s3
import lib.aws.cse as cse

def downloadEncrypted(file_path, bucket_name, s3_key, cse_key_arn):
  body = s3.download(s3_key, bucket_name)
  content = body.read()
  decrypted_content = cse.decrypt(content, cse_key_arn)
  f = open(file_path, 'wb')
  return f.write(decrypted_content)

def download(file_path, bucket_name, s3_key):
  return s3.download_file(s3_key, bucket_name, file_path)
  